### DISCUSSION TITLE
Run command on key press if no window is focused

### DISCUSSION THREAD

**Discussion by exic** [i3/i3#5321] [Q&A] [answered]
It would be nice to start e.g. thunderbird on pressing "t" when I'm on an empty workspace. How is this feasible? Thanks :pray:

**Comment by okraits** [comment, marked as answer]
I think you can do this using the IPC:

https://i3wm.org/docs/ipc.html

* Upon workspace change, check if the workspace is empty
* If yes, then enter a dedicated binding mode: https://i3wm.org/docs/userguide.html#binding_modes
* When a window is opened on that workspace, leave that binding mode

**Reply by Airblader** [reply]
Note that the key will always be bound, i3 doesn't allow binding it conditionally. So you might want to choose something other than just T, unless you don't plan on using that key for anything else (like writing).

**Reply by okraits** [reply]
@Airblader Hmm? When using a binding mode, then the binding for `t` is only active when that binding mode is active. Or am I missing something? There's even an example in the docs for just using `f` and `t`.

**Reply by Airblader** [reply]
I should've read what you suggested better, my bad. Yes, using a binding mode would get around that issue. Sorry for the noise.

**Reply by exic** [reply]
Awesome, got it working. Reused the i3-renameworkspaces.pl that I found somewhere years ago:

```
my $launchmode = 0;
sub start_launchmode {
    my ($data) = @_;
    my @nodes = @{$data->{"current"}->{"nodes"}};
    if (scalar @nodes == 0) { # current workspace has no nodes
        $i3->command("mode launch");
        $launchmode = 1;
    }
}

sub reset_launchmode {
    if ($launchmode == 0) {
        return;
    }
    $i3->command("mode default");
    $launchmode = 0;
}

$i3->subscribe({
    workspace => sub { say('workspace'); updatelabels(); start_launchmode(@_); },
    window    => sub { say('window');    updatelabels(); reset_launchmode(); },
    _error    => sub { say('error');     exit(1);        }
})->recv()->{'success'} or die('Error subscribing to events');
```
